sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/IndexBatchException.java

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.search.documents.models;
+
+import com.azure.core.exception.AzureException;
+import com.azure.search.documents.SearchDocument;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * An {@code IndexBatchException} is thrown whenever Azure Cognitive Search index call was only partially successful.
+ * Users can inspect the indexingResults to determine the operation(s) that have failed.
+ */
+public final class IndexBatchException extends AzureException {
+    private static final long serialVersionUID = -3478124828996650248L;
+    private static final String MESSAGE_FORMAT = "%s of %s indexing actions in the batch failed. The remaining"
+        + " actions succeeded and modified the index. Check indexingResults for the status of each index action.";
+
+    private final ArrayList<IndexingResult> results;
+
+    /**
+     * Constructs an {@code IndexBatchException} from the given {@link IndexDocumentsResult}.
+     *
+     * @param result The DocumentIndexResult returned from the service.
+     */
+    public IndexBatchException(IndexDocumentsResult result) {
+        super(createMessage(result));
+        this.results = new ArrayList<>(result.getResults());
+    }
+
+    /**
+     * Finds all index actions in the given batch that failed and need to be retried, and returns them in a new batch.
+     *
+     * @param originalBatch The batch that partially failed indexing.
+     * @param keyFieldName The name of the key field from the index schema.
+     * @return A new batch containing all the actions from the given batch that failed and should be retried.
+     */
+    public IndexBatchBase<SearchDocument> findFailedActionsToRetry(IndexBatchBase<SearchDocument> originalBatch,
+        String keyFieldName) {
```

Comment:
I'm a little confused on why a `keyFieldName` is needed here, I don't see a key field being used in `IndexBatchBase` or in `IndexAction`.